### PR TITLE
Auth consistency fixes

### PR DESCRIPTION
**Pull request by ahmedhamodi**
Changes:
- implemented a second field within the UserContext that is the field which indicates whether we are logged in with a valid admin user
- this additional field only gets recomputed when auth state changes
- if tokens are still set, no issues (admin user maintains access)
- if tokens are lost, this will fail and show a 403, which in a previous PR I added an okay button which will handle all state issues and move go /login route
- kept "user" as we can use that to pull fields like `displayName` and other info that firebase stores

What to test:
- logical analysis of code for any edge case where we have invalid state
- try to duplicate this when running locally
- try basic login / logout / clear cookies flows

Loom video:
https://www.loom.com/share/fb739486a1a14868a947048ccbdfd6f0

### CODE REVIEW

File: src/Providers/UserProviders.jsx [hunk 1]
```diff
@@ -16,12 +20,21 @@ class UserProvider extends Component {
                 const user = await generateUserDocument(userAuth);
                 this.setState({ user });
             }
+            // TODO: in the future do an authentication route that returns name and email of user?
```
Comment (ahmedhamodi) [inline src/Providers/UserProviders.jsx:23]: If we decide to implement this route in the future, it can also lead to the removal of `user` and the generateUserDocument logic.

However, it is not necessary.

File: src/Components/Admin/admin.js [hunk 2]
```diff
@@ -114,7 +114,7 @@ const useStyles = makeStyles((theme) => ({
 
 export default function Admin() {
     const classes = useStyles();
-    const user = useContext(UserContext);
+    const { user } = useContext(UserContext);
```
Comment (aaronabraham311) [inline src/Components/Admin/admin.js:117]: Just curious, are we able to access the admin's Google Account picture? The designers have put this as a feature, but I wasn't sure if that's possible

File: src/Components/Admin/navbar.js [hunk 3]
```diff
@@ -47,13 +46,15 @@ export default function Navbar({ home }) {
 
     async function handleLogout() {
         handleClose();
-        history.push("/login");
         auth.signOut();
         await httpGet(
             process.env.REACT_APP_ADMIN_BASE_ENDPOINT + "admin_logout"
         );
+        history.push("/login");
```
Comment (aaronabraham311) [inline src/Components/Admin/navbar.js:53]: Why is this reorder needed?

File: src/Providers/UserProviders.jsx [hunk 4]
```diff
@@ -1,10 +1,14 @@
...
+import { httpGet } from "../lib/dataAccess";
+
+const baseEndpoint = process.env.REACT_APP_ADMIN_BASE_ENDPOINT;
 
 export const UserContext = createContext({ user: null });
 class UserProvider extends Component {
     state = {
         user: null,
+        isAdmin: false,
     };
```
Comment (aaronabraham311) [inline src/Providers/UserProviders.jsx:12]: Any plan on refactoring this to functional components in the future?

**Review by aaronabraham311 — APPROVED** [review]
Codewise, I don't see any issues. Just a few questions. I couldn't run this locally because I am way too behind on the repo's changes, so I will leave local testing to @amoliksingh or @jaydhulia

File: src/Components/Admin/navbar.js (hunk 3 above)
Comment (ahmedhamodi) [reply src/Components/Admin/navbar.js:53]: Since the routing is now dependent on isAdmin, which gets recomputed on auth state changes, it is needed to be after the auth.signOut and the resetting of the cookies which the admin_logout does.

File: src/Providers/UserProviders.jsx (hunk 4 above)
Comment (ahmedhamodi) [reply src/Providers/UserProviders.jsx:12]: Hmm, not high priority. Class + functional components work together as far as I know

File: src/Components/Admin/admin.js (hunk 2 above)
Comment (ahmedhamodi) [reply src/Components/Admin/admin.js:121]: I haven't looked into it. Will probably need to do a different ticket for this.

My initial guess is no since you can make a firebase account with a non Google account, so that won't even be an accessible field. And specifically for DCC, they are using outlook emails so they definitely won't have a google account picture.

**Review by jaydhulia — CHANGES_REQUESTED** [review]
Here are the issues I found:
- On the login page (and `/signup` page) itself, when I am not logged in, it makes a request to the backend to `/api/admin/v1/scenarios`, even though an admin is not logged in - this returns a 403 from BE, but this request shouldn't be made in the first place - Loom video: https://www.loom.com/share/4d4c9710134840a0ae10e18ddc494b65
- If an admin logs out, and goes to the `/admin` page (which they may have bookmarked), instead of being shown the login screen, it just shows a completely blank screen. Loom video: https://www.loom.com/share/b8afb9e0f03a49f3ac79d02b1d15b49c
- If an admin is logged in, and they go to the `/login` or `/signup` page, the login form shows up for a second and it just shows a completely blank screen, where you have no option other than to log out. Loom video (same as the second, later on in the video) - https://www.loom.com/share/b8afb9e0f03a49f3ac79d02b1d15b49c

**Comment by ahmedhamodi** [comment]
> * On the login page (and `/signup` page) itself, when I am not logged in, it makes a request to the backend to `/api/admin/v1/scenarios`, even though an admin is not logged in - this returns a 403 from BE, but this request shouldn't be made in the first place

Ideally this makes a call to something like /api/admin/v1/authenticate or something similar. It's the logic we use to see if the cookies are set and are valid. For now it's a /scenarios call. What do you recommend is done instead? No call doesn't make much sense to me because we want to route them to /admin if they are in a valid logged in state when on /login.

> * If an admin logs out, and goes to the `/admin` page (which they may have bookmarked), instead of being shown the login screen, it just shows a completely blank screen.
> * If an admin is logged in, and they go to the `/login` or `/signup` page, the login form shows up for a second and it just shows a completely blank screen, where you have no option other than to log out.

Fixed both.

**Comment by jaydhulia** [comment]
> Ideally this makes a call to something like /api/admin/v1/authenticate or something similar. It's the logic we use to see if the cookies are set and are valid. For now it's a /scenarios call. What do you recommend is done instead? No call doesn't make much sense to me because we want to route them to /admin if they are in a valid logged in state when on /login.

Ah that's what you are making it for. Cool, that's fine for now, we can update this to use a `user_profile` route once we add that to BE routes.

**Comment by ahmedhamodi** [comment]
> Ah that's what you are making it for. Cool, that's fine for now, we can update this to use a `user_profile` route once we add that to BE routes.

Yup! Left a todo in the code for this already

File: src/Providers/UserProviders.jsx [hunk 5]
```diff
@@ -16,12 +20,21 @@ class UserProvider extends Component {
                 const user = await generateUserDocument(userAuth);
                 this.setState({ user });
             }
+            // TODO: in the future do an authentication route that returns name and email of user?
+            try {
+                const isValidUser = await httpGet(baseEndpoint + "scenarios");
+                this.setState({ isAdmin: isValidUser !== undefined });
```
Comment (jaydhulia) [inline src/Providers/UserProviders.jsx:26]: Works for now, in the future we should ideally check if the response code was 403 or not. If it was something like 500, then that doesn't mean the user isn't logged in, just that the backend server is temporarily unavailable.

**Review by jaydhulia — APPROVED** [review]
Added a minor comment, but approving because that can be done in the future. LGTM otherwise.

File: src/Providers/UserProviders.jsx (hunk 5 above)
Comment (ahmedhamodi) [reply src/Providers/UserProviders.jsx:26]: Hmm good point. Will make a mental note of this

Comment (jaydhulia) [reply src/Providers/UserProviders.jsx:26]: I added a PR on the backend, now we should be able to check 200 vs 403 response codes.

**Comment by jaydhulia** [comment]
Backend PR up: https://github.com/uwblueprint/cdc-backend/pull/138, once that is merged, you can make a GET request to `/api/admin/v1/user_profile`

**Review by aaronabraham311 — APPROVED** [review]

**Review by jaydhulia — APPROVED** [review]
